### PR TITLE
feat: Add ngram speculative decoding (Prompt Lookup Decoding, no draft model)

### PR DESCRIPTION
**Pull request by hashkanna**
## Motivation

Partially addresses #192 ("More Speculative algorithms Support -- Call for Contribution"), specifically the ngram algorithms item.  

RFC / design issue: #927. 

This PR implements **Prompt Lookup Decoding (PLD)** -- a trie-based n-gram speculative decoding algorithm that requires **no draft model**. A per-request trie cache is built from previously decoded tokens; BFS queries produce a chain of draft candidates that the target model verifies in a single forward pass via the existing Pallas `verify_tree_greedy` kernel.

This is useful when:
- No draft model is available for the target model
- Memory is constrained (zero additional model memory)
- The generation contains repetitive patterns (e.g., structured output, code, translations)

**Scope:** This PR covers PLD/ngram only. Suffix Decoding and LookAhead Decoding from the same issue bullet are not included and would be separate follow-up PRs.

## Modifications

**New files (3):**
- `python/sgl_jax/srt/speculative/ngram_cache.py` -- Pure Python/NumPy trie-based n-gram cache. Inserts token sequences as n-gram suffixes, queries via BFS ranked by frequency, and produces draft token trees with `retrive_index`, `retrive_next_token`, `retrive_next_sibling` structures compatible with the existing Pallas verification kernel. Supports incremental insertion (`insert_new_suffixes`) to avoid count inflation across decode iterations.
- `python/sgl_jax/srt/speculative/ngram_worker.py` -- `NgramWorker` class that orchestrates draft/verify without a draft model. On decode: queries the trie, allocates KV cache slots, builds a 1D flat custom_mask matching ragged_paged_attention expected layout, runs the target model on the draft tree, and uses `verify_tree_greedy` to accept/reject tokens. `NgramVerifyInput` is registered as a JAX pytree for JIT compatibility. D0 follows the verified_id convention (last known-correct token at tree root).
- `python/sgl_jax/test/speculative/test_ngram_cache.py` -- 33 unit tests covering trie insertion, BFS querying, frequency ranking, breadth limiting, chain/branching tree structures, mask ancestry correctness, incremental insert (no double-counting), padding sentinel handling, custom_mask 1D layout with uneven batch sizes, cycle detection, and `SpeculativeAlgorithm` enum integration. 10 tests are gated on JAX availability.

**Modified files (5):**
- `python/sgl_jax/srt/speculative/spec_info.py` -- Added `NGRAM` to `SpeculativeAlgorithm` enum with `is_ngram()` method and `from_string("NGRAM")` support.
- `python/sgl_jax/srt/server_args.py` -- Added `--speculative-ngram-max-trie-depth` (default 8) and `--speculative-ngram-max-bfs-breadth` (default 1) CLI args, plus `NGRAM` to algorithm choices.
- `python/sgl_jax/srt/managers/scheduler.py` -- Instantiates `NgramWorker` when `speculative_algorithm=NGRAM`, handles ngram `accept_lens`/`allocate_lens` in result processing.
- `python/sgl_jax/srt/managers/schedule_batch.py` -- Guards `EagleDraftInput.ALLOC_LEN_PER_DECODE` access with `is_eagle()` check so ngram decode does not crash on None. Guards `spec_info.allocate_lens` access with None check. Added `reqs` field to `ModelWorkerBatch` so NgramWorker can access request token histories.
- `python/sgl_jax/srt/managers/scheduler_output_processor_mixin.py` -- Added `is_ngram()` to the `output_ids.extend()` branch so ngram-resolved tokens are appended correctly. Ngram frees unaccepted KV slots inside `_forward_decode`, so no additional freeing in the mixin.

## Accuracy Tests

**Unit tests (33/33 pass on TPU v6e-1, JAX 0.8.1, Python 3.12):**
```bash
python3 -m pytest python/sgl_jax/test/speculative/test_ngram_cache.py -v
# 33 passed in 2.03s
```

**End-to-end test on TPU v6e-1 with Qwen/Qwen3-1.7B (bfloat16, TP=1):**
```bash
python -m sgl_jax.launch_server \
  --model-path Qwen/Qwen3-1.7B \
  --trust-remote-code --skip-server-warmup \
  --mem-fraction-static 0.6 --max-running-requests 8 \
  --speculative-algorithm NGRAM \
  --speculative-num-draft-tokens 4 \
  --speculative-ngram-max-trie-depth 8 \
  --speculative-ngram-max-bfs-breadth 1 \
  --disable-overlap-schedule \
  --page-size 64 --attention-backend fa --dtype bfloat16 --tp-size 1
```

Request 1 (`"The capital of France is"`, 32 tokens, temperature=0.0):
```
Paris. The capital of the United States is Washington, D.C. The capital of the
United Kingdom is London. The capital of the Russian Federation is Moscow.
```

Request 2 (`"Write a haiku about machine learning:"`, 64 tokens, temperature=0.0): 64 tokens of coherent output, 40+ decode iterations with no errors.

Server logs show varying acceptance rates:
- Most steps: `accept-len 1.00` (0.25 ratio -- only D0 accepted)
- Two steps: `accept-len 4.00` (1.00 ratio -- all 4 drafts accepted on repeated patterns like "The capital of")
- Two steps: `accept-len 2.00` (0.50 ratio -- partial acceptance)

Existing speculative decoding tests are unaffected (no regressions).

## Benchmarking and Profiling

This is a correctness-first PR. Throughput on v6e-1 was ~0.14-0.55 tok/s due to single chip + no precompile + first-time JIT compilation. Performance optimization is future work (see Limitations below).

## Usage

```bash
python -m sgl_jax.launch_server \
  --model-path <any-model> \
  --speculative-algorithm NGRAM \
  --speculative-num-draft-tokens 4 \
  --speculative-ngram-max-trie-depth 8 \
  --speculative-ngram-max-bfs-breadth 1 \
  --disable-overlap-schedule \
  --attention-backend fa
```

No draft model path required. The ngram cache builds up automatically as tokens are generated.

## Limitations / Future Work

- **Chain mode only** (`--speculative-ngram-max-bfs-breadth=1`). Branching-tree drafts require KV compaction that is not yet implemented; passing breadth > 1 raises `ValueError`.
- **No precompile support** yet -- `run_spec_decode_precompile` is a no-op, so JIT recompilation happens on new batch sizes.
- **Suffix Decoding and LookAhead Decoding** are not included (separate algorithms from the same #192 bullet).
- **`--disable-overlap-schedule` is required** (same as EAGLE).

## Checklist

- [x] Please use English, otherwise it will be closed.
- [x] The purpose of the PR, or link existing issues this PR will resolve.
- [x] The test plan, such as providing test command.
- [ ] (Optional) The necessary documentation update.

### CODE REVIEW

**Comment by JamesBrianD** [comment]
Thanks a lot for your excellent contribution and the careful implementation in this PR! 

Following our project's development process, could you please create a detailed RFC issue to elaborate on the design of this PR, its impact on the existing architecture, and the corresponding future specifications?

We highly appreciate your dedication to SGLang, and looking forward to your follow-up RFC!

**Comment by hashkanna** [comment]
Thanks for the guidance! I’ve filed the RFC issue here: #927

It covers:
- the proposed NGRAM / PLD design
- impact on the existing speculative decoding architecture
- current validated constraints
- future follow-up work and scope boundaries

I’ll also add the RFC link to the PR description for easier reference.
